### PR TITLE
Compare timed UUIDs based on timestamp if possible

### PR DESCRIPTION
**Pull request by sharkwouter**
I've noticed that the UUID.compareTo function does not take the timestamps in timed UUIDs into account in its comparison, which caused some bugs in one of my systems. This PR adds this functionality. Alternatively the javadocs could be improved to explain that timestamps are not used in the comparison if the change in this PR is not wanted.

I've added some checks to the compareTo test which would fail if my change wasn't implemented or implemented correctly. I use 2 UUIDs with the same timestamp and one with a different one.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [ ] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3620/head:pull/3620` \
`$ git checkout pull/3620`

Update a local copy of the PR: \
`$ git checkout pull/3620` \
`$ git pull https://git.openjdk.java.net/jdk pull/3620/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3620`

View PR using the GUI difftool: \
`$ git pr show -t 3620`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3620.diff">https://git.openjdk.java.net/jdk/pull/3620.diff</a>

</details>

### CODE REVIEW

**Comment by sharkwouter** [comment]
The "How to contribute" guide on the OpenJDK website specifies that an issue does not need to be created to be able to contribute and it also does not specify how an issue can be created: https://openjdk.java.net/contribute/

This makes the requirement here to refer to an issue seem really weird. What do I do with this? Can this be skipped or is there a way I can create an account so I can create an issue?

**Comment by sharkwouter** [comment]
Could this be reviewed?

**Comment by sharkwouter** [comment]
Please,someone review this. This solves an actual bug.

File: src/java.base/share/classes/java/util/UUID.java [hunk 1]
```diff
@@ -511,6 +511,12 @@ public boolean equals(Object obj) {
      */
     @Override
     public int compareTo(UUID val) {
+        // Compare the timstamps if both UUIDs are timed UUID
+        // Do not do this if the timestamps are the same
+        if (version() == 1 && val.version() == 1 && timestamp() != val.timestamp()) {
+            return Long.compare(timestamp(), val.timestamp());
+        }
+
```
Comment (plevart) [inline src/java.base/share/classes/java/util/UUID.java:519]: Are you sure this compareTo method is transitive? For example: given 2 version 1 based UUIDs, `a` and `b`, and given the 3rd random UUID `c`, lets say `a < b` and `b < c`. Is it true that `a < c` in all such possible `a`, `b` and `c` ?

Comment (sharkwouter) [reply src/java.base/share/classes/java/util/UUID.java:519]: It should be. If a comparison is done between a timed uuid and a regular one, the timestamp comparison will not be done and the original method will be used.

**Comment by szegedi** [comment]
Comparison of UUIDs is unfortunately fixed in the public API specification. Since the public JavaDoc says "The first of two UUIDs is greater than the second if the most significant field in which the UUIDs differ is greater for the first UUID." that's what the implementation has to be. If you wanted to change this, you'd need to submit a CSR to first change the specification.

Mind you, I'm sympathetic. Even with the current algorithm ordering of UUIDs yields surprising results since the two 64-bit longs inside are compared as signed, so UUIDs starting with 0-7 sort _after_ UUIDs starting with 8-f.

**Comment by sharkwouter** [comment]
Ah, that makes sense then. Is there a guide for how to submit a CSR or is that resevered to people already in the Java development community?

**Comment by szegedi** [comment]
There's a [CSR website](https://wiki.openjdk.java.net/display/csr), but you're right that to submit a CSR request, you must be able to create an issue in JBS, and for that you need to be a committer. What you can do instead is find someone with requisite status in the OpenJDK census and have them sponsor this for you. 

However,  I'm not sure whether such a fundamental change to the way UUIDs are compared would not be accepted. It would also not help you to get it committed into the jdk repo for Java 17 or 18 if the problem also exists on earlier ones; then it'd need to be backported to earlier JDKs too. It's a lot of work that'll take lot of time to show up in a release.

If you're suffering from the lack of desired behavior right now, I think your immediate problem would be better and faster served by implementing a `Comparator<UUID>` and supplying it to the object/method that's comparing or sorting UUIDs. All sorted collections in the JDK and all utility methods that handle `Comparable` objects also can take an explicit `Comparator`.

**Comment by szegedi** [comment]
I'm basing my hunch of this not getting accepted on the fact that the even clearly acknowledged-as-a bug sorting-by-signed-values bug was also rejected from getting fixed, see [JDK-7025832](https://bugs.openjdk.java.net/browse/JDK-7025832).

**Comment by sharkwouter** [comment]
You're probably right, I'll leave this for now.

**Pull request closed without merging**